language/tools/move-bytecode-viewer/src/tui/mod.rs

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+pub mod text_builder;
+pub mod tui_interface;
+
+use crate::tui::tui_interface::TUIInterface;
+use std::{
+    error::Error,
+    io::{self, Stdin, Write},
+};
+use termion::{
+    event::Key,
+    input::{Keys, TermRead},
+};
+use tui::{
+    backend::TermionBackend,
+    layout::{Constraint, Direction, Layout},
+    style::Style,
+    widgets::{Block, Borders, Paragraph},
+    Frame,
+};
+
+pub struct TUI<Interface: TUIInterface> {
+    current_line_number: u16,
+    current_column: u16,
+    interface: Interface,
+    keys: Keys<Stdin>,
+}
+
+impl<Interface: TUIInterface> TUI<Interface> {
+    pub fn new(interface: Interface) -> Self {
+        let keys = io::stdin().keys();
+        Self {
+            current_line_number: 0,
+            current_column: 0,
+            interface,
+            keys,
+        }
+    }
+
+    pub fn redraw<W: Write>(&mut self, f: &mut Frame<TermionBackend<W>>) {
```

Comment:
A lot of fancy fiddly math, maybe comment it just a tad more, but otherwise, thats how these things just tend to be